src/common/transform.h

```diff
@@ -0,0 +1,265 @@
+ *   instance of functor as it's.
+ *
+ * Two form of SegTransform, one with flags, which represents a HostDeviceVector
+ * having one element for each device. Another form is without flags.
+ * Code duplication is a drawback.  If we want to dispatch at runtime, user must
+ * define a `flag` parameter in input functor even if it is not used.
+ */
+template <bool OnDevice = WITH_CUDA>
+struct SegTransform {
+  /*
+   * \brief Segmented transform.
+   *
+   * \tparam Functor Callable object type, accepting spans expanded from
+   *                   varadic HostDeviceVector.
+   * \tparam T       Numeric type of Data.
+   *
+   * \param func     Instance of Functor.
+   * \param range    A Range object specifying begin, end and step (segment size).
+   * \param devices  GPUSet for this function.
+   * \param vectors  Varadic HostDeviceVector<T>, expanded into a set of
+   *                   Span<T> as input for func.
+   */
+  template <typename Functor, typename... T>
+  SegTransform(Functor _func, Range _range, GPUSet _devices,
+               HostDeviceVector<T>*... _vectors) {
+    CHECK((*(_range.end()) - *(_range.begin()) + 1) % _range.GetStep() == 0) <<
+        "Segment length must divides total size.";
+
+    Reshard(_devices, _vectors...);
+
+    bool on_device = _devices != GPUSet::Empty();
+    if (on_device) {
+      LaunchRangeCUDA(_func, _range, _devices, _vectors...);
+    } else {
+      LaunchRangeCPU(_func, _range, UnpackHDV(_vectors)...);
+    }
+  }
+
+  /*
+   * \brief Segmented transform.
+   *
+   * \tparam Functor Callable object type, accepting spans expanded from
+   *                   varadic HostDeviceVector.
+   * \tparam T       Numeric type of Data.
+   *
+   * \param func     Instance of Functor.
+   * \param range    A Range object specifying begin, end and step (segment size).
+   * \param flags    A pointer to HostDeviceVector, with one element for each
+   *                   device.  Passed to func as pointer to element.
+   * \param devices  GPUSet for this function.
+   * \param vectors  Varadic HostDeviceVector<T>, expanded into a set of
+   *                   Span<T> as input for func.
+   */
+  template <typename Functor, typename U, typename... T>
+  SegTransform(Functor _func, Range _range, HostDeviceVector<U>* _flags,
+               GPUSet _devices, HostDeviceVector<T>*... _vectors) {
+    CHECK((*(_range.end()) - *(_range.begin()) + 1) % _range.GetStep() == 0) <<
+        "Segment length must divides total size.";
+
+    Reshard(_devices, _vectors...);
+
+    bool on_device = _devices != GPUSet::Empty();
+    if (on_device) {
+      LaunchRangeCUDA(_func, _range, _flags, _devices, _vectors...);
+    } else {
+      LaunchRangeCPU(_func, _range, _flags->HostPointer(), UnpackHDV(_vectors)...);
+    }
+  }
+
+ private:
+  template <typename... T>
+  void Reshard(GPUSet _devices, HostDeviceVector<T>*... _vectors) {
+    std::vector<HDVAny> vectors {_vectors...};
```

Comment:
Can you just use recursion on the variadic arguments to iterate through the vectors? This would make HDVAny unnecessary.

Reply:
I think it would be possible using std::thread but probably not omp. Not sure how important it is that this function is parallel.

Reply:
I'm not sure either.  That seems to be the right job for OpenMP, one line and worry about nothing.